### PR TITLE
Merge expr bits

### PR DESCRIPTION
**Pull request by martong**
This is an experiment.
Maybe we should copy all `ExprBits`. @balazske what do you think?

### CODE REVIEW

**Comment by balazske** [comment]
It occurs that these values are set by the constructor to default values (at least for `InitListExpr`). Copy of all flags should not cause problems (unless there are differences between imported and original AST that cause different correct settings for the flags but copy is still better than nothing).

File: lib/AST/ASTImporter.cpp [hunk 1]
```diff
@@ -7303,6 +7303,19 @@ Stmt *ASTImporter::Import(Stmt *FromS) {
...
+    auto *FromE = cast<Expr>(FromS);
+    // Copy ExprBitfields, which may not be handled in Expr subclasses
+    // constructors.
+    ToE->setValueKind(FromE->getValueKind());
+    ToE->setObjectKind(FromE->getObjectKind());
+    ToE->setTypeDependent(FromE->isTypeDependent());
+    ToE->setValueDependent(FromE->isValueDependent());
+    ToE->setInstantiationDependent(FromE->isInstantiationDependent());
+    ToE->setContainsUnexpandedParameterPack(
+        FromE->containsUnexpandedParameterPack());
```
Comment (balazske) [inline lib/AST/ASTImporter.cpp:7316]: These flags are stored in an union in `Expr`, but every subclass should use these values (not other things in the same union).

Comment (martong) [reply lib/AST/ASTImporter.cpp:7316]: These flags are stored in the class `ExprBitfields` in `Stmt`. So it is not an union, in fact it is a POD, thus I think it is perfectly safe to overwrite if the node is an `Expr`.
There is a union however in `Stmt`, but that is used amongst the different kind of AST nodes:
```
  union {
    StmtBitfields StmtBits;
    CompoundStmtBitfields CompoundStmtBits;
    IfStmtBitfields IfStmtBits;
    ExprBitfields ExprBits;
    CharacterLiteralBitfields CharacterLiteralBits;
    FloatingLiteralBitfields FloatingLiteralBits;
    UnaryExprOrTypeTraitExprBitfields UnaryExprOrTypeTraitExprBits;
    DeclRefExprBitfields DeclRefExprBits;
    CastExprBitfields CastExprBits;
    CallExprBitfields CallExprBits;
    ExprWithCleanupsBitfields ExprWithCleanupsBits;
    PseudoObjectExprBitfields PseudoObjectExprBits;
    ObjCIndirectCopyRestoreExprBitfields ObjCIndirectCopyRestoreExprBits;
    InitListExprBitfields InitListExprBits;
    TypeTraitExprBitfields TypeTraitExprBits;
    CoawaitExprBitfields CoawaitBits;
  };
```
The members of this union use different storage area than the common `ExprBits`. For example `TypeTraitExprBitfields`, the first member is a padding with the exactly the size of `ExprBits`.
```
  class TypeTraitExprBitfields {
    friend class ASTStmtReader;
    friend class ASTStmtWriter;
    friend class TypeTraitExpr;
    
    unsigned : NumExprBits;
    
    /// \brief The kind of type trait, which is a value of a TypeTrait enumerator.
    unsigned Kind : 8;
    
    /// \brief If this expression is not value-dependent, this indicates whether
    /// the trait evaluated true or false.
    unsigned Value : 1;

    /// \brief The number of arguments to this type trait.
    unsigned NumArgs : 32 - 8 - 1 - NumExprBits;
  };
```
So if an expression is a type trait then it has access to both the common `ExprBits` and the type trait specific bits too.

**Review by balazske — APPROVED** [review]

**Comment by martong** [comment]
https://reviews.llvm.org/D51533

**Pull request merged**